### PR TITLE
add divrem2

### PR DESCRIPTION
**Pull request by hannes14**
an alternative `divrem` based on Singulars `reduce`, to address #3043, #3105

### CODE REVIEW

File: src/ideal/ideal.jl [hunk 1]
```diff
@@ -837,7 +837,26 @@ function divrem(I::sideal{S}, G::sideal{S}; complete_reduction::Bool = false) wh
...
    libSingular.set_option("OPT_REDSB",old_redsb)
    libSingular.set_option("OPT_REDTAIL",old_redtail)
-   return (smodule{S}(R,ptr_T), sideal{S}(R,ptr_Rest), smodule{S}(R,ptr_U))
+   return (sideal{S}(R,ptr_T), sideal{S}(R,ptr_Rest), smodule{S}(R,ptr_U))
+end
+
+@doc raw"""
+    divrem2(I::sideal{S}, G::sideal{S}; complete_reduction::Bool = false) where S <: SPolyUnion
+
+Computes a normalform of the generators of `I` by the generators of `G`,
```
Comment (ederc) [inline src/ideal/ideal.jl:846]:
```suggestion
Computes a normal form of the generators of `I` by the generators of `G`,
```

File: src/ideal/ideal.jl [hunk 2]
```diff
@@ -837,7 +837,26 @@ function divrem(I::sideal{S}, G::sideal{S}; complete_reduction::Bool = false) wh
...
+keeping track of the reduction. Returns a tuple (Quo, Rem, U) where
+`Matrix(I)*Matrix(U) = Matrix(G)*Matrix(Quo) + Matrix(Rem)`
+and `Rem = normalform(I, G)`. `U` is a diagonal matrix of units differing
+from the identity matrix only for local ring orderings.
```
Comment (ederc) [inline src/ideal/ideal.jl:850]:
```suggestion
from the identity matrix only for local monomial orderings.
```

File: src/ideal/ideal.jl [hunk 3]
```diff
@@ -837,7 +837,26 @@ function divrem(I::sideal{S}, G::sideal{S}; complete_reduction::Bool = false) wh
                                                                    false, true, R.ptr)
    libSingular.set_option("OPT_REDSB",old_redsb)
    libSingular.set_option("OPT_REDTAIL",old_redtail)
-   return (smodule{S}(R,ptr_T), sideal{S}(R,ptr_Rest), smodule{S}(R,ptr_U))
+   return (sideal{S}(R,ptr_T), sideal{S}(R,ptr_Rest), smodule{S}(R,ptr_U))
```
Comment (ederc) [inline src/ideal/ideal.jl:840]: Maybe you could also add a test for this change in `divrem`.
Moreover, this is a breaking change for OSCAR's implementation of the `reduce_with_...` functionality.

**Review by ederc — APPROVED** [review]

File: src/ideal/ideal.jl [hunk 4]
```diff
@@ -840,6 +840,25 @@ function divrem(I::sideal{S}, G::sideal{S}; complete_reduction::Bool = false) wh
    return (smodule{S}(R,ptr_T), sideal{S}(R,ptr_Rest), smodule{S}(R,ptr_U))
 end
 
+@doc raw"""
+    divrem2(I::sideal{S}, G::sideal{S}; complete_reduction::Bool = false) where S <: SPolyUnion
+
+Computes a normal form of the generators of `I` by the generators of `G`,
+keeping track of the reduction. Returns a tuple (Quo, Rem, U) where
+`Matrix(I)*Matrix(U) = Matrix(G)*Matrix(Quo) + Matrix(Rem)`
+and `Rem = normalform(I, G)`. `U` is a diagonal matrix of units differing
```
[the remark this answers is not in the document]
Comment (fingolfin) [reply src/ideal/ideal.jl:849]: Maybe `reduce` ?

File: src/ideal/ideal.jl [hunk 5]
```diff
@@ -840,6 +840,25 @@ function divrem(I::sideal{S}, G::sideal{S}; complete_reduction::Bool = false) wh
    return (smodule{S}(R,ptr_T), sideal{S}(R,ptr_Rest), smodule{S}(R,ptr_U))
 end
 
+@doc raw"""
+    divrem2(I::sideal{S}, G::sideal{S}; complete_reduction::Bool = false) where S <: SPolyUnion
+
+Computes a normal form of the generators of `I` by the generators of `G`,
+keeping track of the reduction. Returns a tuple (Quo, Rem, U) where
```
Comment (fingolfin) [inline src/ideal/ideal.jl:847]: The docstring for `divrem` does not include "keeping track of the reduction", is that an intentional difference, and if so, what does it signify?

File: src/ideal/ideal.jl (hunk 4 above)
[the remark this answers is not in the document]
Comment (fingolfin) [reply src/ideal/ideal.jl:849]: Guessing here:
```suggestion
and `Rem = reduce(I, G; complete_reduction)`. `U` is a diagonal matrix of units differing
```

File: test/ideal/sideal-test.jl [hunk 6]
```diff
@@ -713,4 +713,6 @@ end
    @test Singular.Matrix(a)*Singular.Matrix(u) == Singular.Matrix(b)*Singular.Matrix(q)+Singular.Matrix(r)
    @test gens(reduce(a, b, complete_reduction=false))[1] ==y^3+x*y
    @test gens(reduce(a, b, complete_reduction=true))[1] == y^3
+   q2,r2,u2 = divrem2(a, b)
+   @test Singular.Matrix(a)*Singular.Matrix(u2) == Singular.Matrix(b)*Singular.Matrix(q2)+Singular.Matrix(r2)
```
[the remark this answers is not in the document]
Comment (hannes14) [reply test/ideal/sideal-test.jl:717]: Yes, reduce is the name of the function in Singular.jl, normalform is used in textbooks.
additional requirement: for global ordeirngs: u2 is a unit matrix of n x n, (n being the number of generators of a)

File: src/ideal/ideal.jl (hunk 5 above)
Comment (hannes14) [reply src/ideal/ideal.jl:847]: divrem uses internally lift and Rem may by different from reduce(I,G) (if G is not a Groebner basis),
while divrem2 uses internally reduce and Rem is reduce(I,G)

**Pull request merged**